### PR TITLE
feat: add quarto cheat sheet extension with cheat sheet option

### PR DESCRIPTION
**Pull request by Revathyvenugopal162**
- Update the cheatsheet option
- Adapt to new quarto cheatsheet format

### CODE REVIEW

File: src/ansys_sphinx_theme/__init__.py [hunk 1]
```diff
@@ -446,7 +448,11 @@ def configure_theme_logo(app: Sphinx):
     if not theme_options.get("logo"):
         theme_options["logo"] = pyansys_logo
 
-    if not isinstance(theme_options.get("logo"), str):
+    if isinstance(theme_options.get("logo"), str) and theme_options.get("logo") not in [
+        "ansys",
+        "pyansys",
+        "no_logo",
+    ]:
```
Comment (jorgepiloto) [inline src/ansys_sphinx_theme/__init__.py:455]: Shouldn't this be part of a separate PR?

File: src/ansys_sphinx_theme/__init__.py [hunk 2]
```diff
@@ -464,6 +470,105 @@ def configure_theme_logo(app: Sphinx):
...
+        Path to the output directory.
+    output_png : str
+        Name of the output PNG file.
+    """
+    try:
+        doc = pymupdf.open(pdf_path)  # open document
+        for page in doc:
+            pix = page.get_pixmap()
+            pix.save(os.path.join(output_dir, output_png))
+    except Exception as e:
```
Comment (jorgepiloto) [inline src/ansys_sphinx_theme/__init__.py:491]: Let's avoid capturing generic exceptions. Can we narrow which exceptions are involved here?

File: src/ansys_sphinx_theme/__init__.py (hunk 1 above)
Comment (Revathyvenugopal162) [reply src/ansys_sphinx_theme/__init__.py:455]: > Shouldn't this be part of a separate PR?

yes, when i was testing this PR with pymechanical, it was blocking the build. i will revert this change and open another PR

File: src/ansys_sphinx_theme/__init__.py (hunk 2 above)
Comment (Revathyvenugopal162) [reply src/ansys_sphinx_theme/__init__.py:491]: Sure, adding that.

Comment (Revathyvenugopal162) [reply src/ansys_sphinx_theme/__init__.py]: added in [1544c20](https://github.com/ansys/ansys-sphinx-theme/pull/428/commits/1544c20529ed0969baca26555a6471a42bda43cd)

**Comment by Revathyvenugopal162** [comment]
Testing with https://github.com/ansys/pymechanical/pull/845

File: src/ansys_sphinx_theme/__init__.py [hunk 3]
```diff
@@ -463,6 +378,137 @@ def configure_theme_logo(app: Sphinx):
...
+    file_path = cheatsheet_file.parent
+    output_dir_path = pathlib.Path(app.outdir) / output_dir
+    try:
+        # Add the cheatsheet to the Quarto project
+        result = subprocess.run(
+            ["quarto", "add", "ansys/pyansys-quarto-cheatsheet@v1", "--no-prompt"],
+            cwd=file_path,
+            capture_output=True,
+            text=True,
+        )
```
Comment (jorgepiloto) [inline src/ansys_sphinx_theme/__init__.py:461]: Let's isolate the version within the scope of this function:

```suggestion
        VERSION = "v1"
        result = subprocess.run(
            ["quarto", "add", f"ansys/pyansys-quarto-cheatsheet@{VERSION}", "--no-prompt"],
            cwd=file_path,
            capture_output=True,
            text=True,
        )
```

File: src/ansys_sphinx_theme/__init__.py [hunk 4]
```diff
@@ -463,6 +378,137 @@ def configure_theme_logo(app: Sphinx):
...
+            file_path = cheatsheet_file.parent / file
+            if file_path.exists():
+                file_path.unlink()
+
+            # if static folder is clean, delete it
+        if not list(cheatsheet_file.parent.glob("_static/*")):
+            cheatsheet_file.parent.joinpath("_static").rmdir()
+
+    except subprocess.CalledProcessError as e:
+        raise RuntimeError(f"Failed to build Quarto cheatsheet: {e}, ensure Quarto is installed.")
```
Comment (jorgepiloto) [inline src/ansys_sphinx_theme/__init__.py:503]:
```suggestion
    except subprocess.CalledProcessError as e:
        raise RuntimeError(f"Failed to build Quarto cheatsheet: {e}. Ensure Quarto is installed.")
```

**Review by jorgepiloto — CHANGES_REQUESTED** [review]

File: src/ansys_sphinx_theme/__init__.py [hunk 5]
```diff
@@ -464,6 +470,105 @@ def configure_theme_logo(app: Sphinx):
...
+    app : sphinx.application.Sphinx
+        Application instance for rendering the documentation.
+    """
+    cheatsheet_options = app.config.html_theme_options.get("cheatsheet", {})
+
+    if not cheatsheet_options:
+        return
+
+    cheatsheet_file = cheatsheet_options.get("file", "")
+    output_dir = cheatsheet_options.get("output_dir", "")
```
[the remark this answers is not in the document]
Comment (Revathyvenugopal162) [reply src/ansys_sphinx_theme/__init__.py]: removed in [82cb447](https://github.com/ansys/ansys-sphinx-theme/pull/428/commits/82cb4473ce6bf1179be5a487690b351e0a2c8953)

File: src/ansys_sphinx_theme/__init__.py [hunk 6]
```diff
@@ -463,6 +378,138 @@ def configure_theme_logo(app: Sphinx):
...
+    output_dir : pathlib.Path
+        Path to the output directory.
+    output_png : str
+        Name of the output PNG file.
+    """
+    try:
+        doc = pymupdf.open(pdf_path)  # open document
+        for page in doc:
+            pix = page.get_pixmap()
+            pix.save(os.path.join(output_dir, output_png))
```
Comment (jorgepiloto) [inline src/ansys_sphinx_theme/__init__.py:398]: Please, avoid using `os.path`. Use `pathlib` instead:

```suggestion
            pix.save(output_dir / output_png)
```

File: src/ansys_sphinx_theme/__init__.py [hunk 7]
```diff
@@ -463,6 +378,138 @@ def configure_theme_logo(app: Sphinx):
...
+            "_static/ansys.png",
+        ]
+        for file in supplementary_files:
+            file_path = cheatsheet_file.parent / file
+            if file_path.exists():
+                file_path.unlink()
+
+            # if static folder is clean, delete it
+        if not list(cheatsheet_file.parent.glob("_static/*")):
+            cheatsheet_file.parent.joinpath("_static").rmdir()
```
Comment (jorgepiloto) [inline src/ansys_sphinx_theme/__init__.py:501]:
```suggestion
        # If static folder is clean, delete it
        if not list(cheatsheet_file.parent.glob("_static/*")):
            cheatsheet_file.parent.joinpath("_static").rmdir()
```

**Review by jorgepiloto — CHANGES_REQUESTED** [review]
Almost there. Some minor suggestions and we will merge this.

File: src/ansys_sphinx_theme/__init__.py [hunk 8]
```diff
@@ -24,10 +24,11 @@
 import logging
 import os
 import pathlib
+import subprocess
 from typing import Any, Dict
 
 from docutils.nodes import document
-import requests
+import pymupdf
```
Comment (RobPasMue) [inline src/ansys_sphinx_theme/__init__.py:31]: Have we verified that we can use this pacakge? It's license is GNU AFFERO GPL 3.0. I'm guessing there is no problem but I am also interested in its approval for other libraries.

**Review by RobPasMue — CHANGES_REQUESTED** [review]

File: src/ansys_sphinx_theme/__init__.py (hunk 8 above)
Comment (RobPasMue) [reply src/ansys_sphinx_theme/__init__.py:31]: https://www.gnu.org/licenses/agpl-3.0.html

File: src/ansys_sphinx_theme/__init__.py [hunk 9]
```diff
@@ -463,6 +378,138 @@ def configure_theme_logo(app: Sphinx):
...
+    if not cheatsheet_file:
+        return
+
+    cheatsheet_file = pathlib.Path(app.srcdir) / cheatsheet_file
+    file_name = str(cheatsheet_file.name)
+    file_path = cheatsheet_file.parent
+    output_dir_path = pathlib.Path(app.outdir) / output_dir
+    try:
+        # Add the cheatsheet to the Quarto project
+        VERSION = "v1"
```
Comment (SMoraisAnsys) [inline src/ansys_sphinx_theme/__init__.py:456]: I would move that at the beginning of the file and rename it, e.g. `CHEAT_SHEET_VERSION`

Comment (Revathyvenugopal162) [reply src/ansys_sphinx_theme/__init__.py:456]: sure, moving it

File: src/ansys_sphinx_theme/__init__.py (hunk 8 above)
Comment (jorgepiloto) [reply src/ansys_sphinx_theme/__init__.py:31]: I do not think this is an issue for other libraries since they use the theme for development. However, it is an issue within this project for building the wheelhouse.

Comment (RobPasMue) [reply src/ansys_sphinx_theme/__init__.py:31]: Yep - that was my concern. Redistributing is not well seen with AGPL 3.0 from waht I understood, and the wheelhouse is a clear redistribution of software...

Comment (Revathyvenugopal162) [reply src/ansys_sphinx_theme/__init__.py]: Thank you for raising the issue, I initially used this approach to avoid adding extra dependencies on ImageMagick or Poppler. However, given the issues we encountered, I reverted to using pdf2image and added a note explaining the change. Please review

File: doc/source/user-guide/options.rst [hunk 10]
```diff
@@ -238,23 +232,14 @@ main ``index.rst`` file and the ``learning.rst`` file in its "Getting started" s
...
-    If you set ``"needs_download": True", you should provide the ``html_static_path`` option
-    in the ``conf.py`` file to specify the location for downloading the cheat sheet to.
-
-    For example, if you want to download the cheat sheet to the ``_build/html/_static/`` directory,
-    you should add the following line in the ``conf.py`` file:
-
-    .. code-block:: pycon
-
-        html_static_path = ["_static"]
+    To use this feature, you must have the `quarto <https://quarto.org/>` package installed. To create thumbnails of generated pdf files,
```
Comment (Revathyvenugopal162) [inline doc/source/user-guide/options.rst:243]:
```suggestion
    To use this feature, you must have the `quarto <https://quarto.org/>` package installed. To create thumbnails of generated PDF files,
```

File: src/ansys_sphinx_theme/__init__.py [hunk 11]
```diff
@@ -395,12 +395,12 @@ def convert_pdf_to_png(pdf_path: pathlib.Path, output_dir: pathlib.Path, output_
         Name of the output PNG file.
     """
     try:
-        doc = pymupdf.open(pdf_path)  # open document
-        for page in doc:
-            pix = page.get_pixmap()
-            pix.save(output_dir / output_png)
-    except RuntimeError as e:
-        raise RuntimeError(f"Failed to convert PDF to PNG: {e}")
+        images = convert_from_path(pdf_path, 500)
```
Comment (RobPasMue) [inline src/ansys_sphinx_theme/__init__.py:398]: Can we do the import in the case we are interested on calling this function only? I am expecting that, since system dependencies are required, this might lead to failure on resolving imports even and not only when calling it. Maybe I'm wrong but I would constraint the import call to this function only

**Review by RobPasMue — CHANGES_REQUESTED** [review]

File: src/ansys_sphinx_theme/__init__.py (hunk 11 above)
Comment (Revathyvenugopal162) [reply src/ansys_sphinx_theme/__init__.py:398]: yes,i will add that

**Comment by Revathyvenugopal162** [comment]
tested with https://github.com/ansys/pymechanical/pull/845
